### PR TITLE
JAPI-440 Process valid TLKs - currently dropping "empty" TLKs

### PR DESCRIPTION
**Pull request by jpmcmu**
- Changed how partition processor creates the cap record in TLKs
Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Review by drealeed — APPROVED** [review]
Changes make sense to me

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java [hunk 1]
```diff
@@ -213,9 +219,25 @@ public static Object getMaxValueForType(FieldType type) throws Exception
...
+                        {
+                            maxVal.concat(maxStringValue);
+                            i -= maxStringValue.length();
+                        }
+                        return maxVal;
+                    }
+                    else
+                    {
+                        return maxStringValue;
+                    }
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java:240]: The fieldtype enum defines other numerics, boolean and other complex types, have we confirmed they're invalid types for index?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java [hunk 2]
```diff
@@ -260,20 +260,35 @@ else if (fd.getFieldType() == FieldType.STRING)
...
-                log.warn("No TLK partition provided to partition processor; all partitions will be returned for all filters.");
-                return;
-            }
-            else
+            // This would represent some form of corruption or an issue during the read
+            throw new Exception("Mismatch between # of data partitions and records in TLK: " + tlkRecords.size() + " # partitions: " + this.dataPartitions.length);
+        }
+
+        // Create a max value cap record
+        if (tlkRecords.size() > 0)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java:270]: is it ok to fall through if size() == 0 ?

**Review by rpastrana — COMMENTED** [review]
@jpmcmu a few comments, mostly minor. However, I didn't see a corresponding junit. Can we generate data to exercise this new behavior?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java (hunk 2 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java:270]: Yep

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java (hunk 1 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java:240]: Correct, there are limited types that can be created in the key of the index

**Comment by jpmcmu** [comment]
@rpastrana pushed up code review changes, please review

**Comment by rpastrana** [comment]
@jpmcmu minor, but javadoc is reporting warnings for the exception:

 Warning:  Javadoc Warnings
Warning:  /home/runner/work/hpcc4j/hpcc4j/dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java:204: warning: no description for @throws
Warning:  * @throws Exception
Warning:  ^
Warning:  /home/runner/work/hpcc4j/hpcc4j/dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java:217: warning: no description for @throws
Warning:  * @throws Exception
Warning:  ^

**Comment by jpmcmu** [comment]
@rpastrana javadoc fixed & squashed

**Review by rpastrana — APPROVED** [review]

**Pull request merged**